Add chain mapper implementation

jonathonwalz opened the pull request:
I've written a potential implementation of a chain mapper (such as discussed in #58).

I've been using this to successfully use both PHPCR and ORM for a few months now, but only for editing existing content. I've only recently started looking into creating content with the createjs. 

I made a breaking change to the RdfMapperInterface to handle resolving object names. Since I don't have a very in-depth knowledge of the workings of createphp I'm not sure how much this will affect users of the library. (Or, for that matter, most of my changes. I've only spent a cursory amount of time looking at create)

flack commented:
Hi!

thanks for the PR! I'm a bit busy at the moment, so reviewing it properly might take a moment. Just a quick question: What's the rationale for removing the `canonicalName` function? I'm not using it myself a lot, but I remember that @dbu needed it for his integration. 

@dbu: could you take a look at this and let us know if you see any problem resulting from this in your CMF integration?

jonathonwalz commented:
I removed it since I needed `objectToName` on the mapper. The chain mapper needs the object to determine which mapper to use. `canonicalName` just takes a classname so it won't be able to resolve that to a specific mapper. So I moved `objectToName` to the mapper class - at which point I wasn't sure what the point of having `canonicalName` was anymore so I removed it.

dbu commented:
great! i will have a look as soon as i find time. pretty busy on this
end too.

dbu commented on src/Midgard/CreatePHP/Mapper/BaseDoctrineRdfMapper.php:
```diff
@@ -58,9 +61,10 @@ public function store(EntityInterface $entity)
     {
-        $refl = new \ReflectionClass($className);
+        $refl = new \ReflectionClass($object);
+        $className = $refl->getName();
```
as we have the object now, we should not use reflection but instead simply do this (no need even for an "if")

    \Doctrine\Common\Util\ClassUtils::getClass($object);

(and lets add a use statement for Doctrine\Common\Util\ClassUtils)

dbu commented on src/Midgard/CreatePHP/Metadata/AbstractRdfDriver.php:
```diff
@@ -72,19 +72,6 @@ protected function buildInformation($child, $identifier, $field, &$add_default_v
-     *
-     * The default implementation uses get_class on objects
-     */
-    public function objectToName($object, RdfMapperInterface $mapper)
```
i agree with moving this to the mapper.

but i suggest we add an `@deprected` to this method and simply call $mapper->objectToName , rather than doing the BC break now.

dbu commented on src/Midgard/CreatePHP/Metadata/RdfDriverInterface.php:
```diff
@@ -34,15 +34,6 @@
-     *
-     * @return string the canonical name of this object
-     */
-    public function objectToName($object, RdfMapperInterface $mapper);
```
same for this, lets just add a `@deprecated` for now and not remove it yet.

dbu commented on src/Midgard/CreatePHP/RdfChainableMapperInterface.php:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Map from CreatePHP to your domain objects
+ *
+ * You can have a mapper per type or a generic mapper that handles all types.
```
please update this description to explain what the chainable mapper is. the general description of what a mapper is can stay in RdfMapperInterface

dbu commented on src/Midgard/CreatePHP/Mapper/AbstractRdfMapper.php:
```diff
@@ -145,14 +156,17 @@ protected function getField($object, RdfElementDefinitionInterface $child)
-     * @return string exactly the same as $className
+     * The default implementation uses get_class on objects
      */
-    public function canonicalName($className)
```
lets just `@deprecate` this method and have objectToName still call canonicalName so that we keep BC.

dbu commented on src/Midgard/CreatePHP/Mapper/BaseDoctrineRdfMapper.php:
```diff
@@ -58,9 +61,10 @@ public function store(EntityInterface $entity)
      *
      * use getRealClass if className names a doctrine proxy class.
      */
-    public function canonicalName($className)
```
for BC, lets just keep this method as it was but add the new objectToName method.

we can then eventually go for a version 2.0 of this library, but for now lets keep things working.

dbu commented on src/Midgard/CreatePHP/RdfMapperInterface.php:
```diff
@@ -112,6 +95,25 @@ public function store(EntityInterface $entity);
+     *
+     * @return string the canonical name of this object
+     */
+    public function objectToName($object);
```
this is the one actual BC break. i suggest to bump branch-alias in composer.json (which we seem to have missed to bump to 1.0) to 1.1-dev and add a note for this interface change in the changelog. something like:

* add 1.0 title to current changelog
* above, add 1.1 title
* say "Added ChainRdfMapper and ChainedRdfMapperInterface to allow using more than one mapper in parallel."... and "BC break: If you implemented your own mapper, note that RdfMapperInterface::objectToName was added"...

dbu commented:
i just tested editing in the cmf with this branch, and it works fine. (i did not test the chain mapper, just whether the phpcr-odm mapper still works correctly). good job @jonathonwalz !

i added some ideas to mostly keep backwards compatibility and a few cleanups. otherwise i think this is fine. a bit of documentation would certainly help, just mentioning that this even exists and explaining the idea a little bit.

@jonathonwalz do you use this in the context of symfony? would you mind doing a PR on CreateBundle to allow configuring the orm mapper and make the bundle use the chain mapper, once this is merged? would be great!

flack commented:
thanks @dbu for the review! I agree with your suggestions for `deprecated` annotations, and can make the `branch-alias` changes once the PR gets updated

dbu commented:
the branch-alias could also be changed in the PR, thats the easiest way.

flack commented:
@dbu: Sure, I was just offering to do something so that I don't look too lazy :-)

jonathonwalz commented:
I've made the changes to just deprecate canonical name. 

I do use it with symfony. I'd be happy to do a PR on the CreateBundle too.

dbu commented on src/Midgard/CreatePHP/Mapper/ChainRdfMapper.php:
```diff
@@ -0,0 +1,183 @@
+    protected function getMapperForObject($object)
+    {
+        foreach ($this->mappers as $mapper)
+        {
```
if we follow symfony codestyle, this `{` would need to be at the end of the line above.

dbu commented on src/Midgard/CreatePHP/Mapper/ChainRdfMapper.php:
```diff
@@ -0,0 +1,183 @@
+            }
+        }
+
+        throw new RuntimeException("No mapper can create a subject for object.");
```
lets do
`sprintf('None of the registered mappers can create the subject for object of class %s', get_class($object))`

dbu commented:
great, i think this is good. i just noted a few codestyle issues. not sure how much @flack as the maintainer of this libs cares about them - maybe @flack you can also just merge this and then run the [codestyle fixer](https://github.com/fabpot/PHP-CS-Fixer) over the codebase.

dbu commented:
and looking forward to a PR on the bundle. maybe lets discuss how this can look in https://github.com/symfony-cmf/CreateBundle/issues/109 so that we don't spam flack with symfony specific discussion :-)

jonathonwalz commented:
I fixed the code style for the lines that I touched and updated the exception message.

dbu commented:
great! i am now +1 for merging without further changes. 

@flack can you tell me if you need any input? i suggest we wait with tagging 1.1 until we see things work properly in CreateBundle so that we can still tweak things here without creating confusion.

flack commented:
Alright, merged now. Thanks @jonathonwalz for the code and @dbu for the review. Sorry that I'm not very responsive right now, but I'm on a very tight deadline for another project...

@dbu If you want to, I can give you collaborator access to the repo. You know the code as well as I do, and I guess at the moment you guys even use it more than me :-)

dbu commented:
thanks a lot for merging.

i won't mind having merge rights if you want to give me them. we should
agree how to make sure that i know you saw a PR and think its ok if i
merge it eventually. or do you want to fully trust me to judge on that?
i would still ping you when i am not fully convinced

flack commented:
Alright, I've added you as collaborator now. For merging stuff, I think we can to a simple ACK/NACK system. If I don't respond in time, it'll be my own fault :-)

dbu commented:
thanks!
